tests/test_ad.py

```diff
@@ -62,25 +62,28 @@ def getRootDSE(self, request, reply):
 
 class AbstractLdapActiveDirectoryTestCase:
     @defer.inlineCallbacks
-    def setUp(self):
+    def make_ldap_server(self, api):
         self.ldap_server = yield ensureDeferred(
             create_ldap_server(_ActiveDirectoryLDAPServer)
         )
+
+        self.auth_provider = create_auth_provider(
+            self.ldap_server,
+            api,
+            config=self.getConfig(),
+        )
+
+    def make_mock(self, side_effects):
         module_api = Mock(
             spec_set=[
                 "check_user_exists",
                 "get_qualified_user_id",
                 "register_password_auth_provider_callbacks",
             ]
         )
-        module_api.check_user_exists.return_value = make_awaitable(True)
+        module_api.check_user_exists.side_effect = side_effects
```

Comment:
```suggestion
        module_api.check_user_exists.side_effect = lambda user_id: user_id.lower()
```
I'd be tempted to write this and not make the caller pass this in.

*If a test wants to override this, they still could — but AFAICT most of the tests actually don't care about overriding it*

Reply:
So I am slightly confused by the `.lower()` in this suggestion-the actual sql being run is `sql = "SELECT name, password_hash FROM users WHERE lower(name) = lower(?)"` which should return the non-lowercased name, correct? If this function lowers the case of the of the username then the logic in this PR is incorrect as it will have the same problem we are trying to fix wherein a access token will then be created based on the lowercase version of the name and won't work in subsequent api calls.

Reply:
This is just to give a meaningful response in the test whilst also being quite concise about it.
You could also write `lambda user_id: make_awaitable(user_id)` and assume that the correct user ID is always the one that the 'user' passes in instead; it's just a choice rather than being a matter of wrong and right.

The job of `check_user_exists` is to return the 'correct' user ID; with this suggestion applied then the default behaviour in tests will be to lowercase user IDs within tests and treat those as canonical. (If you imagine that you can only register lowercase user IDs, like today, then that leads to the same behaviour as the query you mention.)

You're right that the query returns the non-lowercased name if that's how it's written in the database: that's because in that case, the correct user ID doesn't have lowercase in it.

The idea with this suggested change is that any test which cared about / wanted to test that we could correct usernames back in to their uppercase form would overwrite this mock with something more relevant for what it was doing.

Reply:
Okay great thanks for explaining! I just wanted to make sure I didn't get my wires crossed somewhere. Appreciate it.